### ISSUE TITLE
Incorrect results in System.Numerics.BigInteger.Log()

### ISSUE THREAD

**Issue by speshuric** [dotnet/corefx#4965]
System.Numerics.BigInteger.Log() on very large BigInteger calculate incorrect results.
Repro:

    BigInteger b = BigInteger.One << 64 << Int32.MaxValue;
    Double fact = BigInteger.Log(b, 2);
    Double expect = ((double)Int32.MaxValue + 64);
    Console.WriteLine("Fact: {0}", fact);
    Console.WriteLine("Expected: {0}", expect);

Expected output:
```
Fact: 2147483711
Expected: 2147483711
```

Fact:
```
Fact: -2147483585
Expected: 2147483711
```

This caused by "overflow" in Log() method (line with `--->`):
```
            // measure the exact bit count
            int c = NumericsHelpers.CbitHighZero((uint)h);
--->        int b = value._bits.Length * 32 - c;

            // extract most significant bits
            ulong x = (h << 32 + c) | (m << c) | (l >> 32 - c);

            // let v = value, b = bit count, x = v/2^b-64
            // log ( v/2^b-64 * 2^b-64 ) = log ( x ) + log ( 2^b-64 )
            return Math.Log(x, baseValue) + (b - 64) / Math.Log(baseValue, 2);
```
when `value._bits.Length > Int32.MaxValue / 32` then `b` is negative and return value is incorrect.

It can be easy to fix. Change
```
int b = value._bits.Length * 32 - c;
```
to
```
long b = (long)value._bits.Length * 32 - c;
```
Whole method is quite heavy, so one addition conversion doesn't affect performance.
I did it in d77d57d402839f1da8aed30c2231f30c6f7b92bb
Test for this case added with 2d20c59dcf85269a06d928a475d039646cf5e066

If this is OK I will send Pull Request

**Comment by speshuric** [comment]
I just discovered that problem not only in `Log`.

Same pattern found in `public static explicit operator Double(BigInteger value)`. Repro:
```
                BigInteger temp = BigInteger.One << Int32.MaxValue << Int32.MaxValue;
                double d = (double)temp;
```
Expected `d`: positive infinity, in fact: 0,25
Posiible fix: to check `bits.Length` if it greater than `some_threshold` (or less than `some_threshold`) return positive or negative infinity. This when `some_threshold` can be found from maximum exponent in double. 

I can write fix and tests for this case.

But there is not the end of "story". Note that in our case we operate super-large values `(2^(2^32))` and more. This values cannot be stored as strings and built from strings. Is it correct to allow operate such values or we sould restrict them? What is the right way?

@axelheer can you look this issue? I noticed that current implementation of `Log` and many other operations are made by you (and in previous implementation incorrect results starts from less numbers)

**Comment by axelheer** [comment]
Thanks for the analysis, your fix for `Log` looks good to me. I think we can fix `operator Double` the same way: just using `long` beforehand and then a `min(exp, int.maxvalue)` for `GetDoubleFromParts`, which should lead to infinity looking at its implementation (or just returning infinity if `exp` grows too large...).

What do you mean with "not the end of story"? I did a quick search for code which multiplies `Length` and that are the only two places looking bad to me. I think as long as we don't get rubbish as result (and we get that for the two situations you pointed out), we should not limit the size of big integers. But note: I'm not maintainer of this project, just my two cents.

And can you rephrase "in previous implementation incorrect results starts from less numbers" for me either, I simply don't understand what you mean. :confused:

**Comment by axelheer** [comment]
cc: @KrzysztofCwalina @stephentoub

**Comment by mellinoe** [comment]
Thanks for the detailed analysis and repo, @speshuric . It seems like you already have a fix and a test written, would you care to open a PR with those?

> What do you mean with "not the end of story"? I did a quick search for code which multiplies Length and that are the only two places looking bad to me.

I think he means that the indices for the bits of the massive number won't even fit in the array, because we're limited to Int32 array index values (Array.Length => Int32). Is that what you meant, @speshuric ?

**Comment by speshuric** [comment]
@axelheer yes, I see that `GetDoubleFromParts` works with infinity, but it seemed to me that if `bits` too long then manipulation with `man` is unnecessory.
@axelheer before your improvement incorrect results for `Log` was for
```
BigInteger b = BigInteger.One << 2 << Int32.MaxValue;
double fact = BigInteger.Log(b, 2);
```
You can check it in 4.5. After improvement result are correct up to `BigInteger.One << 64 << Int32.MaxValue` (wich is first incorrect)

@mellinoe yes, fixes are written, tests for `Double()` didn't complete, but it will not take a long time. I'll open PR in a few days.
@mellinoe , @axelheer : "not the end of story" I mean that there is other restrictions in `BigInteger`: 
* ctor `BigInteger(byte[])` and function `ToByteArray()` limited near 2GB
* without `gcAllowVeryLargeObjects` all arrays limited to 2GB
* duration of `ToString()` is at the least quadratic function of length. Even for `Int32.MaxValue` bits it's impossible to get result
* `Parse`/`TryParse` do not work with huge numbers
* If every 10 bits generates ~3 decimal digits then it will be 48 bits in 2-byte encodings. _If_ we can make decimal representation of number with `Int32.MaxValue` bits it will be about 1,5 G string

And note that those limitations are "by design". So I think that numbers more than `Int32.MaxValue` bits are quite dangerous and may be explicit limitation to `Int32.MaxValue` bits will be better than attempts to support it. 

But anyway the first and short step is to fix existing obvious bugs. Explicit size limit is only suggestion for future.

**Comment by speshuric** [comment]
> I'll open PR in a few days.

Few days appeared to be a bit shorter. I've opened PR.
In `double()` operator checks for infinity placed before high dwords manipulation. `Log` fixed as in initial message. 
Let me know if the tests are too heavy or should cover more use cases

**Comment by axelheer** [comment]
@speshuric thanks for the additional explanation!
